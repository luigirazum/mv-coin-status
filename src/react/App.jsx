import { Routes, Route } from 'react-router-dom';
import Home from './routes/Home';
import CoinsLayout from './routes/CoinsLayout';
import AllCoins from './routes/AllCoins';
import CoinDetails from './routes/CoinDetails';
import '../styles/App.css';

const App = () => (
  <Routes>
    <Route index element={<Home />} />
    <Route path="coins" caseSensitive element={<CoinsLayout />}>
      <Route index caseSensitive element={<AllCoins />} />
      <Route path=":id" caseSensitive element={<CoinDetails />} />
    </Route>
    <Route path="*" element={<h3>ERROR</h3>} />
  </Routes>
);

export default App;
