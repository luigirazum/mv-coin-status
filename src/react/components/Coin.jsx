import { useSelector } from 'react-redux';
import PropTypes from 'prop-types';
import { selectCoinById } from '../../redux/coins/coinsSelectors';

const Coin = ({ id }) => {
  const coin = useSelector((store) => selectCoinById(store, id));
  const {
    icon, name, symbol, rank,
  } = coin;

  return (
    <article className="coinItem">
      <section className="itemSection itemHeader">
        <h3 className="itemName">{name}</h3>
        <div className="iconContainer">
          <img src={icon} className="coinIcon" alt={name} />
        </div>
      </section>
      <section className="itemSection">
        <div className="itemDescription">
          <h3 className="itemCategory rankCategory">Rank</h3>
          <p className="itemData rankData">{rank}</p>
        </div>
        <div className="itemDescription">
          <h4 className="itemCategory">Symbol</h4>
          <p className="itemData">{symbol}</p>
        </div>
      </section>
    </article>
  );
};

Coin.propTypes = {
  id: PropTypes.string.isRequired,
};

export default Coin;
