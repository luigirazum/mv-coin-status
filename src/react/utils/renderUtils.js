import PropTypes from 'prop-types';
import React from 'react';
import {
  BrowserRouter, MemoryRouter, Routes, Route,
} from 'react-router-dom';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

// import your setupStore as a basic setup
import setupStore from '../../redux/store';

/** ------------------------------------------------------------
 *   > to test components that require a store/state/router
 *  ------------------------------------------------------------ */
const renderWithProvider = (
  ui,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    ...renderOptions
  } = {},
  {
    routerWrapper = null,
    routes = ['/'],
    path = '/',
    user = userEvent,
  } = {},
) => {
  const RouterWrapper = ({ children }) => {
    if (routerWrapper.memory) {
      return (
        <MemoryRouter initialEntries={routes}>
          <Routes>
            <Route path={path} element={children} />
          </Routes>
        </MemoryRouter>
      );
    }

    return (
      <BrowserRouter>
        <Routes>
          <Route path={path} element={children} />
        </Routes>
      </BrowserRouter>
    );
  };

  RouterWrapper.propTypes = {
    children: PropTypes.element.isRequired,
  };

  const Wrapper = ({ children }) => (
    <Provider store={store}>
      {
      routerWrapper
        ? <RouterWrapper>{children}</RouterWrapper>
        : children
      }
    </Provider>
  );

  Wrapper.propTypes = {
    children: PropTypes.element.isRequired,
  };

  const component = Wrapper({ children: ui });
  if (routerWrapper && routerWrapper.browser) window.history.pushState({}, 'test page', routes[0]);

  // Return an object with the store and all of RTL's query functions
  return ({
    store,
    component,
    user,
    ...render(ui, { wrapper: Wrapper, ...renderOptions }),
  });
};

export default renderWithProvider;
