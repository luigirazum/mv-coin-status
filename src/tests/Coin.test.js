import { screen } from '@testing-library/react';
import renderer, { format as prettyFormat } from 'react-test-renderer';
import { renderWithProviders } from '../react/utils/renderUtils';
import { mockCoins } from './mocks/mockData';
import Coin from '../react/components/Coin';

let container = null;
const [bitCoin] = mockCoins.coins;

beforeEach(() => {
  // setup a DOM element as a render target
  container = renderWithProviders(<Coin id={bitCoin.id} />, {
    preloadedState: {
      coins: {
        assets: [mockCoins.coins],
        status: { type: 'idle' },
        filter: { active: false, by: '' },
        error: null,
      },
    },
  }).container;
  document.body.appendChild(container);
});

afterEach(() => {
  //  cleanup on exiting
  // unmountComponentAtNode(container);
  container.remove();
  container = null;
});

/**
 * START: Coin component test
 */
describe('Coin component tests', () => {
  /**
   * START: Coin render
  */
  describe('Coin must be rendered', () => {
    test('should render a name', () => {
      expect(screen.getByTestId('coinName')).toBeInTheDocument();
    });
    test('should render a rank', () => {
      expect(screen.getByTestId('coinRank')).toBeInTheDocument();
    });
    test('should render a symbol', () => {
      expect(screen.getByTestId('coinSymbol')).toBeInTheDocument();
    });
    test('should render a logo', () => {
      expect(screen.getByTestId('coinLogo')).toBeInTheDocument();
    });
  });
  /**
   * END: Coin render
  */

  describe('Coin component snapshot test', () => {
    test('snapshot renders correctly', () => {
      const renderedWithProviders = renderWithProviders(<Coin id={bitCoin.id} />, {
        preloadedState: {
          coins: {
            assets: [mockCoins.coins],
            status: { type: 'idle' },
            filter: { active: false, by: '' },
            error: null,
          },
        },
      });

      container = renderedWithProviders.container;
      const { component } = renderedWithProviders;

      const tree = renderer.create(component);
      const jsonTree = tree.toJSON();
      expect(jsonTree).toMatchSnapshot();
    });
  });
});
/**
 * END: Coin component tests
*/
