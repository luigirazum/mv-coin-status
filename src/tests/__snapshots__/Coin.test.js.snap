// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Coin component tests Coin component snapshot test snapshot renders correctly 1`] = `
<div>
  <article
    class="coinItem"
    data-testid="bitcoin"
  >
    <section
      class="itemSection itemHeader"
    >
      <h3
        class="itemName"
        data-testid="coinName"
      />
      <div
        class="iconContainer"
      >
        <img
          class="coinIcon"
          data-testid="coinLogo"
        />
      </div>
    </section>
    <section
      class="itemSection"
    >
      <div
        class="itemDescription"
      >
        <h3
          class="itemCategory rankCategory"
        >
          Rank
        </h3>
        <p
          class="itemData rankData"
          data-testid="coinRank"
        />
      </div>
      <div
        class="itemDescription"
      >
        <h4
          class="itemCategory"
        >
          Symbol
        </h4>
        <p
          class="itemData"
          data-testid="coinSymbol"
        />
      </div>
    </section>
  </article>
</div>
`;
